src/mono/mono/mini/mini-llvm.c

```diff
@@ -1462,7 +1462,12 @@ emit_volatile_store (EmitContext *ctx, int vreg)
 
 	if (var && var->flags & (MONO_INST_VOLATILE|MONO_INST_INDIRECT)) {
 		g_assert (ctx->addresses [vreg]);
+#ifdef HOST_WASM
+		/* Need volatile stores otherwise the compiler might move them */
+		mono_llvm_build_store (ctx->builder, convert (ctx, ctx->values [vreg], type_to_llvm_type (ctx, var->inst_vtype)), ctx->addresses [vreg], TRUE, LLVM_BARRIER_NONE);
+#else
```

Comment:
Out of curiosity, why is this wasm-specific?

Reply:
On wasm, we use INDIRECT to force reference variables to the stack, so we need those stores to not be optimized.

Reply:
Tbh, not sure why we don't do this on all platfoms, at least for VOLATILE. It didn't seem to cause problems on other platforms, just on wasm.